File: features/lead/src/main/java/com/nivasafinance/features/lead/controller/LeadNotesController.java
package com.nivasafinance.features.lead.controller;

import com.nivasafinance.common.base.model.PaginatedResponse;
import com.nivasafinance.common.base.model.PaginationRequest;
import com.nivasafinance.common.constants.ApiConstants;
import com.nivasafinance.common.annotations.RequirePermission;
import com.nivasafinance.features.lead.dto.LeadNoteCreateRequest;
import com.nivasafinance.features.lead.dto.LeadNoteCreateResponse;
import com.nivasafinance.features.lead.dto.LeadNoteResponse;
import com.nivasafinance.features.lead.dto.LeadNoteUpdateRequest;
import com.nivasafinance.features.lead.service.LeadNoteReadService;
import com.nivasafinance.features.lead.service.LeadNoteWriteService;
import jakarta.validation.Valid;
import lombok.AllArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.UUID;

@RestController
@RequestMapping(ApiConstants.V1 + "/leads/{leadId}/notes")
@AllArgsConstructor
public class LeadNotesController {

    private final LeadNoteWriteService leadNoteWriteService;
    private final LeadNoteReadService leadNoteReadService;

    @PostMapping
    @RequirePermission(permissionName = "CREATE_LEAD_NOTES")
    public ResponseEntity<LeadNoteCreateResponse> createLeadNote(
            @PathVariable UUID leadId,
            @Valid @RequestBody LeadNoteCreateRequest request
    ) {
        LeadNoteCreateResponse response = leadNoteWriteService.createLeadNote(leadId, request);
        return ResponseEntity.status(HttpStatus.CREATED).body(response);
    }

    @GetMapping
    @RequirePermission(permissionName = "READ_LEAD_NOTES")
    public ResponseEntity<PaginatedResponse<LeadNoteResponse>> getAllLeadNotes(
            @PathVariable UUID leadId,
            @Valid PaginationRequest paginationRequest) {
        
        PaginatedResponse<LeadNoteResponse> notes = leadNoteReadService.getAllLeadNotes(
                leadId, paginationRequest);
        return ResponseEntity.ok(notes);
    }

    @GetMapping("/{noteId}")
    @RequirePermission(permissionName = "READ_LEAD_NOTES")
    public ResponseEntity<LeadNoteResponse> getLeadNoteById(
            @PathVariable UUID leadId,
            @PathVariable UUID noteId) {
        LeadNoteResponse note = leadNoteReadService.getLeadNoteById(leadId, noteId);
        return ResponseEntity.ok(note);
    }

    @PutMapping("/{noteId}")
    @RequirePermission(permissionName = "UPDATE_LEAD_NOTES")
    public ResponseEntity<Void> updateLeadNote(
            @PathVariable UUID leadId,
            @PathVariable UUID noteId,
            @Valid @RequestBody LeadNoteUpdateRequest request) {
        leadNoteWriteService.updateLeadNote(leadId, noteId, request);
        return ResponseEntity.noContent().build();
    }

    @DeleteMapping("/{noteId}")
    @RequirePermission(permissionName = "DELETE_LEAD_NOTES")
    public ResponseEntity<Void> deleteLeadNote(
            @PathVariable UUID leadId,
            @PathVariable UUID noteId) {
        leadNoteWriteService.deleteLeadNote(leadId, noteId);
        return ResponseEntity.noContent().build();
    }
}

File: features/lead/src/main/java/com/nivasafinance/features/lead/controller/LeadNotesController.kt
package com.nivasafinance.features.lead.controller

import base.model.PaginatedResponse
import base.model.PaginationRequest
import com.nivasafinance.features.lead.dto.LeadNotesResponse
import com.nivasafinance.features.lead.service.LeadNotesService
import com.nivasafinance.features.notes.dto.NotesRequest
import com.nivasafinance.features.notes.dto.NotesUpdateRequest
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.util.*

@RestController
@RequestMapping("/api/leads")
class LeadNotesController(
    private val leadNotesService: LeadNotesService
) {

    // @GetMapping("/notes/all")
    // fun getAllNotes(
    //     @RequestParam(defaultValue = "0") offset: Int,
    //     @RequestParam(defaultValue = "20") limit: Int,
    //     @RequestParam(defaultValue = "createdAt") sortBy: String,
    //     @RequestParam(defaultValue = "ASC") sortDirection: String
    // ): ResponseEntity<PaginatedResponse<List<LeadNotesResponse>>> {
    //     val paginationRequest = PaginationRequest(
    //         offset = offset,
    //         limit = limit,
    //         sortBy = sortBy,
    //         sortDirection = sortDirection
    //     )
    //     val notes = leadNotesService.getAllNotes(paginationRequest)
    //     return ResponseEntity.ok(notes)
    // }

    @PostMapping("/{leadId}/notes")
    fun addNotesToLead(
        @PathVariable leadId: UUID,
        @RequestParam taskId: UUID? = null,
        @RequestBody addNotesRequest: NotesRequest
    ): ResponseEntity<LeadNotesResponse> {
        val leadNotes = leadNotesService.addNotesToLead(leadId, taskId, addNotesRequest)
        return ResponseEntity.status(HttpStatus.CREATED).body(leadNotes)
    }

    @GetMapping("/{leadId}/notes/all")
    fun getNotesByLeadId(
        @PathVariable leadId: UUID,
        @RequestParam taskId: UUID? = null,
        @RequestParam(defaultValue = "0") offset: Int,
        @RequestParam(defaultValue = "20") limit: Int,
        @RequestParam(defaultValue = "createdAt") sortBy: String,
        @RequestParam(defaultValue = "ASC") sortDirection: String
    ): ResponseEntity<PaginatedResponse<List<LeadNotesResponse>>> {
        val paginationRequest = PaginationRequest(
            offset = offset,
            limit = limit,
            sortBy = sortBy,
            sortDirection = sortDirection
        )

        val leadNotes = leadNotesService.getTaskNotes(leadId, taskId, paginationRequest)
        return ResponseEntity.ok(leadNotes)
    }

    @PatchMapping("/{leadId}/notes/{notesId}")
    fun patchNotesByLeadId(
        @PathVariable leadId: UUID,
        @RequestParam taskId: UUID? = null,
        @PathVariable notesId: UUID,
        @RequestBody updateNotesRequest: NotesUpdateRequest
    ): ResponseEntity<LeadNotesResponse> {
        val updatedNotes = leadNotesService.patchNotesById(leadId, taskId, notesId, updateNotesRequest)
        return ResponseEntity.ok(updatedNotes)
    }

    @DeleteMapping("/{leadId}/notes/{notesId}")
    fun deleteNotesByLeadId(
        @PathVariable leadId: UUID,
        @RequestParam taskId: UUID? = null,
        @PathVariable notesId: UUID
    ): ResponseEntity<Unit> {
        leadNotesService.deleteNotesById(leadId, taskId, notesId)
        return ResponseEntity.noContent().build()
    }
}
